BUG: random: Generator.integers(2**32) always returned 0.

WarrenWeckesser opened the pull request:
When the input to Generator.integers was `2**32`, the value `2**32-1`
was being passed as the `rng` argument to the 32-bit Lemire method,
but that method requires `rng` be strictly less than `2**32-1`.

The fix was to handle `2**32-1` by calling next_uint32 directly.
This also works for the legacy code without changing the stream
of random integers from `randint`.

Closes gh-16066.

<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

WarrenWeckesser committed:
BUG: random: Generator.integers(2**32) always returned 0.

When the input to Generator.integers was 2**32, the value 2**32-1
was being passed as the `rng` argument to the 32-bit Lemire method,
but that method requires `rng` be strictly less then 2**32-1.

The fix was to handle 2**32-1 by calling next_uint32 directly.
This also works for the legacy code without changing the stream
of random integers from `randint`.

Closes gh-16066.

seberg commented:
Any chance of the same issue for `2**64` (not through `integers` maybe)?

seberg commented:
@bashtage would you have time to review this to make sure its good?

seberg reviewed (approved):
LGTM, I think fine to put in as is. It seems to me the test should include also drawing a single sample to cover both paths, though?
Otherwise, will leave it in case Kevin wants to have a look, but lets not hesitate to merge to get 1.18.4 out.

WarrenWeckesser commented:
@seberg wrote
>It seems to me the test should include also drawing a single sample to cover both paths, though?

The test in there now is a crude regression test.  I'll replace that with a repeatability test for `2**32-1`, `2**32` and `2**32+1`.

seberg commented:
Right, just need to also cover the path where only a single number is drawn then. The approach is good in any case, I was wondering whether it makes sense to put an assert for this into the Lemirs method. We could just do that, although I doubt it helps much with flushing these issues out.

WarrenWeckesser committed:
MAINT: random: Add assert() statements.

Assert that an invalid value (2**n-1 for n = 8, 16, 32, 64) has not
been passed to the Lemire function.

WarrenWeckesser commented:
It took a few tries, but the Travis-CI tests on the `s390x` platform finally ran successfully.

I've changed the unit test to a set of repeatability tests for the values `2**32-1`, `2**32` and `2**32+1`. For each value, a test is run with `size=7` and `size=None`.

I also added `assert` calls in the C implementations of the Lemire method to check for the invalid value.

bashtage commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -1450,22 +1469,34 @@ void random_bounded_uint64_fill(bitgen_t *bitgen_state, uint64_t off,
       for (i = 0; i < cnt; i++) {
-        out[i] = off + buffered_bounded_masked_uint32(bitgen_state, rng, mask,
-                                                      &bcnt, &buf);
+        out[i] = off + (uint64_t) next_uint32(bitgen_state);
```
Is this code used in `RandomState`, and does this preserve the stream guarantee?  RS uses only masked, so intercepting before this might have been called could can an issue. 

Would it be safest to added a test in RandomState too with the same end points?

WarrenWeckesser replied:
This is tested by the test that was added in gh-14501: https://github.com/numpy/numpy/blob/20f10762fc20a29923b6943235c2577b8e3ac9f3/numpy/random/tests/test_randomstate_regression.py#L178

bashtage commented on numpy/random/src/distributions/distributions.c:
```diff
@@ -1337,6 +1348,14 @@ uint64_t random_bounded_uint64(bitgen_t *bitgen_state, uint64_t off,
+       * call next_uint32 directly.  This also works when use_masked is True,
+       * so we handle both cases here.
+       */
+      return off + (uint64_t) next_uint32(bitgen_state);
```
I think the same would apply here as well.

WarrenWeckesser replied:
Yes, this needs a test.  When calling `randint` with scalar values of `low` and `high`, this path is not reached, even when `size` is `None`.    It turns out that the function `random_bounded_uint64` is called by the code that handles broadcasting for `random.randint`.  I'll add a repeatability test for that.

bashtage reviewed (changes requested):
LGTM, modulo checking that the stream in `RandomState` is not affected.

WarrenWeckesser committed:
TST: random: Add more repeatability tests for random integers.

Add repeatability tests for when the range of the integers is `2**32`
(and `2**32 +/- 1` for good measure) with broadcasting.  The underlying
functions called by Generator.integers and random.randint when the
inputs are broadcast are different than when the inputs are scalars.

bashtage reviewed (approved):
Extra test was all I noticed.  It has been added.

seberg commented:
@WarrenWeckesser cool, thanks for the nice tests. Unfortunately due to 32bit longs, the tests are failing on windows (and presumably 32bit linux).

WarrenWeckesser commented:
@seberg, yup, just noticed that. Looking into it.

WarrenWeckesser committed:
TST: random: Skip a test if integers are 32 bit.

WarrenWeckesser commented:
For the new `randint` test, I used the same `skipif` decorator as in https://github.com/numpy/numpy/pull/14501.  Tests are all green now.

seberg commented:
Seems good to just skip. Thanks for tracking it down and fixing it quickly Warren!

charris commented:
AFAIK, RandomState does not use the Lemire method, only the mask method. For repeatability tests, I used fairly long streams and hashed them, then checked the hash. Takes up a lot less room. 

Repeatability is not guaranteed for the Generator methods. At some point we should have correctness checks, maybe as a separate set of tests.

WarrenWeckesser commented:
@charris wrote
> Repeatability is not guaranteed for the Generator methods.

My understanding is that we have repeatability tests for  the Generator methods so we can detect a change, and create a release note about the change if turns out that the change is unavoidable.

WarrenWeckesser commented:
>At some point we should have correctness checks, maybe as a separate set of tests.

Agreed.  The issue gh-15911 that you created looks like a good place to discuss this.

iyanmv commented:
Was this ever merged to 1.17.x? So, correct me if I'm wrong, but the first version of NumPy that can be used without issues [gh-16066](https://github.com/numpy/numpy/issues/16066) and [gh-14774](https://github.com/numpy/numpy/issues/14774) is 1.18.4, right?

bashtage commented:
The last 17 branch was released before this PR was merged, so it was never in the 17 branch.  It was backported to 18 and in 19.0.

iyanmv commented:
Okay, so I guess a bump in requirements-min.txt is needed in my case.
